Add context menu directive

azrikahar opened the pull request:
Fixes #9926

Ref: https://github.com/directus/directus/pull/9271#discussion_r740579195

---

Might have butchered this implementation since I've never personally added custom directives, but hope it's in the right direction 😅

azrikahar committed:
Add context menu directive

rijkvanzanten commented on app/src/directives/context-menu/context-menu.ts:
```diff
@@ -0,0 +1,40 @@
+import { Directive, DirectiveBinding } from 'vue';
+
+function mounted(element: HTMLElement, binding: DirectiveBinding): void {
+	const contextMenu = binding.instance?.$refs['contextMenu'];
```
I don't think we can make this automated 🤔 There's gonna be cases where theres's more than one context menu in the same parent component. I think we should update this so the contextMenu ref is pulled in from the binding, which then means the use becomes:

`<element v-context-menu="contextMenuRef">`

azrikahar replied:
> There's gonna be cases where theres's more than one context menu in the same parent component. 

That's true 😄 Updated the PR with this change.

rijkvanzanten commented:
Other than that binding, this is looking good!

azrikahar committed:
require specifying a ref name

joselcvarela committed:
Merge remote-tracking branch 'origin/main' into context-menu-directive

joselcvarela committed:
fix empty context menu
@azrikahar

azrikahar commented:
@joselcvarela thanks for fixing the empty context menu. That's a tricky one, assuming there's more items added to the list so the condition will need to be updated 🤔 but it'll have to do for now I guess.

joselcvarela committed:
Merge remote-tracking branch 'origin/main' into context-menu-directive

joselcvarela committed:
listen contextmenu even if no menu attached
In some cases we have an area where would like to not open the browser context menu. For example, an area where the context menu does not have any options because of conditions like permissions

joselcvarela committed:
only 'Show hidden...' if hidden collections exists

joselcvarela committed:
fix arrow chunk overlapping items on v-menu

joselcvarela commented:
Hey @azrikahar 
Just made a few more changes, check it out.
Basically:
- if `v-contextmenu` is attached, always listen to context menu in order to prevent browser context menu. This is useful when we want to attach a context menu but the `v-menu` has a condition.
- only `Show hidden collections` if there is hidden collections. The fix above is used here (because `v-menu` will not render)
- the arrow on `v-menu` was annoying me for a while because it was on top of items in `v-menu` so a little part was shown. Not anymore!!! 😅

rijkvanzanten commented:
> the arrow on v-menu was annoying me for a while because it was on top of items in v-menu so a little part was shown. Not anymore!!! 😅

I'm sure @benhaynes will appreciate this

rijkvanzanten committed:
Merge branch 'main' into context-menu-directive

benhaynes commented:
You know it!! Thanks, @joselcvarela!! ❤️
